src/graph-update-scanner.service.ts

```diff
@@ -0,0 +1,95 @@
+import { Injectable, Logger, OnApplicationBootstrap, OnApplicationShutdown } from '@nestjs/common';
+import { BlockHash } from '@polkadot/types/interfaces';
+import { InjectQueue } from '@nestjs/bullmq';
+import { Queue } from 'bullmq';
+import { SchedulerRegistry } from '@nestjs/schedule';
+import { MILLISECONDS_PER_SECOND, SECONDS_PER_MINUTE } from 'time-constants';
+import { ConfigService } from './config/config.service';
+import { UpdateTransitiveGraphs, createGraphUpdateJob } from './interfaces/graph-update-job.interface';
+import { BlockchainService } from './blockchain/blockchain.service';
+import { BlockchainScannerService } from './blockchain-scanner.service';
+import { ReconnectionCacheMgrService } from './cache/reconnection-cache-mgr.service';
+import { ReconnectionServiceConstants } from './constants';
+
+@Injectable()
+export class GraphUpdateScannerService extends BlockchainScannerService implements OnApplicationBootstrap, OnApplicationShutdown {
+  private readonly QUEUE_HIGH_WATER_MARK: number;
+
+  async onApplicationBootstrap() {
+    // Set up recurring interval
+    const interval = setInterval(() => this.scan(), this.configService.getBlockchainScanIntervalMinutes() * SECONDS_PER_MINUTE * MILLISECONDS_PER_SECOND);
+    this.schedulerRegistry.addInterval('blockchainScan', interval);
+
+    // Kick off initial scan
+    const initialTimeout = setTimeout(() => this.scan(), 0);
+    this.schedulerRegistry.addTimeout('initialScan', initialTimeout);
+    await super.onApplicationBootstrap();
+  }
+
+  async onApplicationShutdown(signal?: string | undefined) {
+    if (this.schedulerRegistry.doesExist('timeout', 'initialScan')) {
+      this.schedulerRegistry.deleteTimeout('initialScan');
+    }
+
+    if (this.schedulerRegistry.doesExist('interval', 'blockchainScan')) {
+      this.schedulerRegistry.deleteInterval('blockchainScan');
+    }
+
+    super.onApplicationShutdown(signal);
+  }
+
+  constructor(
+    cacheManager: ReconnectionCacheMgrService,
+    @InjectQueue(ReconnectionServiceConstants.GRAPH_UPDATE_QUEUE_NAME) private graphUpdateQueue: Queue,
+    private readonly configService: ConfigService,
+    private schedulerRegistry: SchedulerRegistry,
+    blockchainService: BlockchainService,
+  ) {
+    super(cacheManager, blockchainService, new Logger(GraphUpdateScannerService.name));
+    this.QUEUE_HIGH_WATER_MARK = Number(this.configService.getQueueHighWater());
+  }
+
+  protected async checkInitialScanParameters(): Promise<boolean> {
+    const queueSize = await this.graphUpdateQueue.count();
+    if (queueSize > 0) {
+      this.logger.log('Deferring next blockchain scan until queue is empty');
+      return false;
+    }
+
+    return true;
```

Comment:
nit: does checkInitialScanParameters() is meant to see if any pending items in update queue? If so don't add more but wait

Just for my understanding

Reply:
Yes--there are 2 queue-related controls for this scanner:
1. High-water mark: scan will stop once there are N items pending on the queue
2. Empty queue--scan can resume once existing queue has been drained